File: api/app/routes/documents.py
import uuid
from pathlib import Path

from fastapi import APIRouter, BackgroundTasks, Depends, File, HTTPException, Request, UploadFile, status
from fastapi.responses import FileResponse
from slowapi import Limiter
from slowapi.util import get_remote_address
from sqlalchemy.orm import Session

from app import metrics
from app.database import SessionLocal, get_db
from app.logger import get_logger
from app.models import Document
from app.security import require_session
from app.settings import settings

log = get_logger("carver.documents")
_limiter = Limiter(key_func=get_remote_address)
router = APIRouter(prefix="/documents", tags=["documents"])

VALID_DOC_TYPES = {"cv", "references", "passport", "stcw", "eng1", "photo"}
ALLOWED_EXTENSIONS = {".pdf", ".doc", ".docx", ".jpg", ".jpeg", ".png"}
ALLOWED_MIME_TYPES = {
    "application/pdf",
    "application/msword",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "image/jpeg",
    "image/png",
}


def _upload_dir() -> Path:
    path = settings.UPLOAD_DIR
    path.mkdir(parents=True, exist_ok=True)
    return path


def _get_doc(db: Session, user_key: str, doc_type: str) -> Document | None:
    return (
        db.query(Document)
        .filter(Document.user_key == user_key, Document.doc_type == doc_type)
        .first()
    )


def _scan_document_background(doc_id: int, file_path: Path, doc_type: str) -> None:
    """Run AI document scan in a background task and persist the result."""
    from app.services.document_scanner import scan_document

    try:
        summary = scan_document(file_path, doc_type)
    except Exception:
        log.exception("Background document scan failed | doc_id=%d | doc_type=%s", doc_id, doc_type)
        return

    if not summary:
        return

    db = SessionLocal()
    try:
        doc = db.query(Document).filter(Document.id == doc_id).first()
        if doc:
            doc.scanned_text = summary
            db.commit()
            log.info("Document scanned | doc_id=%d | doc_type=%s | len=%d", doc_id, doc_type, len(summary))
    except Exception:
        log.exception("Failed to persist scan result | doc_id=%d", doc_id)
        db.rollback()
    finally:
        db.close()


@router.get("")
def list_documents(
    session: dict = Depends(require_session),
    db: Session = Depends(get_db),
):
    user_key = session["sub"]
    docs = db.query(Document).filter(Document.user_key == user_key).all()
    return {
        d.doc_type: {
            "original_name": d.original_name,
            "uploaded_at": d.created_at.isoformat(),
            "scanned": d.scanned_text is not None,
        }
        for d in docs
    }


@router.post("/{doc_type}", status_code=status.HTTP_200_OK)
@_limiter.limit("20/minute")
async def upload_document(
    request: Request,
    doc_type: str,
    background_tasks: BackgroundTasks,
    file: UploadFile = File(...),
    session: dict = Depends(require_session),
    db: Session = Depends(get_db),
):
    if doc_type not in VALID_DOC_TYPES:
        raise HTTPException(status_code=400, detail=f"Invalid doc_type. Must be one of: {', '.join(VALID_DOC_TYPES)}")

    suffix = Path(file.filename or "").suffix.lower()
    if suffix not in ALLOWED_EXTENSIONS:
        raise HTTPException(status_code=400, detail=f"File type not allowed. Accepted: {', '.join(ALLOWED_EXTENSIONS)}")

    if file.content_type and file.content_type not in ALLOWED_MIME_TYPES:
        raise HTTPException(status_code=400, detail="MIME type not allowed.")

    contents = await file.read()
    if len(contents) > settings.UPLOAD_MAX_BYTES:
        raise HTTPException(status_code=413, detail=f"File exceeds {settings.UPLOAD_MAX_BYTES // (1024*1024)} MB limit.")
    if len(contents) == 0:
        raise HTTPException(status_code=400, detail="Empty file.")

    user_key = session["sub"]
    upload_dir = _upload_dir()

    existing = _get_doc(db, user_key, doc_type)
    if existing:
        old_file = upload_dir / existing.stored_name
        old_file.unlink(missing_ok=True)
        existing.original_name = file.filename or "upload"
        existing.stored_name = f"{uuid.uuid4().hex}{suffix}"
        existing.scanned_text = None
        (upload_dir / existing.stored_name).write_bytes(contents)
        db.commit()
        db.refresh(existing)
        metrics.increment("doc_uploads")
        log.info("Document replaced | user=%s | doc_type=%s | size=%d", user_key, doc_type, len(contents))
        background_tasks.add_task(_scan_document_background, existing.id, upload_dir / existing.stored_name, doc_type)
        return {"ok": True, "original_name": existing.original_name}

    stored_name = f"{uuid.uuid4().hex}{suffix}"
    (upload_dir / stored_name).write_bytes(contents)
    doc = Document(
        user_key=user_key,
        doc_type=doc_type,
        original_name=file.filename or "upload",
        stored_name=stored_name,
    )
    db.add(doc)
    db.commit()
    db.refresh(doc)
    metrics.increment("doc_uploads")
    log.info("Document uploaded | user=%s | doc_type=%s | size=%d", user_key, doc_type, len(contents))
    background_tasks.add_task(_scan_document_background, doc.id, upload_dir / stored_name, doc_type)
    return {"ok": True, "original_name": doc.original_name}


@router.delete("/{doc_type}", status_code=status.HTTP_200_OK)
@_limiter.limit("20/minute")
def delete_document(
    request: Request,
    doc_type: str,
    session: dict = Depends(require_session),
    db: Session = Depends(get_db),
):
    if doc_type not in VALID_DOC_TYPES:
        raise HTTPException(status_code=400, detail="Invalid doc_type.")

    user_key = session["sub"]
    doc = _get_doc(db, user_key, doc_type)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found.")

    upload_dir = _upload_dir()
    (upload_dir / doc.stored_name).unlink(missing_ok=True)
    db.delete(doc)
    db.commit()
    metrics.increment("doc_deletes")
    log.info("Document deleted | user=%s | doc_type=%s", user_key, doc_type)
    return {"ok": True}


@router.get("/{doc_type}/file")
def download_document(
    doc_type: str,
    session: dict = Depends(require_session),
    db: Session = Depends(get_db),
):
    if doc_type not in VALID_DOC_TYPES:
        raise HTTPException(status_code=400, detail="Invalid doc_type.")

    user_key = session["sub"]
    doc = _get_doc(db, user_key, doc_type)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found.")

    upload_dir = _upload_dir()
    file_path = (upload_dir / doc.stored_name).resolve()
    if not str(file_path).startswith(str(upload_dir.resolve())):
        raise HTTPException(status_code=400, detail="Invalid file path.")
    if not file_path.exists():
        raise HTTPException(status_code=404, detail="File missing on server.")

    metrics.increment("doc_downloads")
    log.info("Document downloaded | user=%s | doc_type=%s", user_key, doc_type)
    return FileResponse(
        path=str(file_path),
        filename=doc.original_name,
        media_type="application/octet-stream",
    )

File: jobcarver-go/internal/handlers/documents/documents.go
// Package documents ports api/app/routes/documents.py — upload / list / delete /
// download of crew documents. AI document scanning is out of scope for the Go
// port (see TODO below); everything else (validation, storage, metadata) is kept.
package documents

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"jobcarver/internal/config"
	"jobcarver/internal/db"
	"jobcarver/internal/httpx"
	"jobcarver/internal/metrics"
	"jobcarver/internal/models"
	"jobcarver/internal/security"
)

var validDocTypes = map[string]bool{
	"cv": true, "references": true, "passport": true, "stcw": true, "eng1": true, "photo": true,
}

var allowedExtensions = map[string]bool{
	".pdf": true, ".doc": true, ".docx": true, ".jpg": true, ".jpeg": true, ".png": true,
}

var allowedMimeTypes = map[string]bool{
	"application/pdf": true,
	"application/msword": true,
	"application/vnd.openxmlformats-officedocument.wordprocessingml.document": true,
	"image/jpeg": true,
	"image/png":  true,
}

// Register mounts /documents (all routes require a session).
func Register(r chi.Router) {
	r.Route("/documents", func(r chi.Router) {
		r.Use(security.RequireSession)
		r.Get("/", list)
		r.Post("/{docType}", upload)
		r.Delete("/{docType}", remove)
		r.Get("/{docType}/file", download)
	})
}

func uploadDir() string {
	dir := config.S.UploadDir
	_ = os.MkdirAll(dir, 0o755)
	return dir
}

func getDoc(userKey, docType string) (models.Document, bool) {
	var doc models.Document
	if err := db.DB.Where("user_key = ? AND doc_type = ?", userKey, docType).First(&doc).Error; err != nil {
		return models.Document{}, false
	}
	return doc, true
}

func list(w http.ResponseWriter, r *http.Request) {
	sess := security.SessionFromContext(r.Context())
	var docs []models.Document
	db.DB.Where("user_key = ?", sess.Sub).Find(&docs)
	out := map[string]any{}
	for _, d := range docs {
		out[d.DocType] = map[string]any{
			"original_name": d.OriginalName,
			"uploaded_at":   d.CreatedAt.Format(time.RFC3339),
			"scanned":       d.ScannedText != nil,
		}
	}
	httpx.JSON(w, http.StatusOK, out)
}

func upload(w http.ResponseWriter, r *http.Request) {
	sess := security.SessionFromContext(r.Context())
	userKey := sess.Sub
	docType := chi.URLParam(r, "docType")

	if !validDocTypes[docType] {
		httpx.WriteError(w, http.StatusBadRequest,
			"Invalid doc_type. Must be one of: cv, references, passport, stcw, eng1, photo", "")
		return
	}

	file, header, err := r.FormFile("file")
	if err != nil {
		httpx.WriteError(w, http.StatusBadRequest, "Missing file.", "")
		return
	}
	defer file.Close()

	suffix := strings.ToLower(filepath.Ext(header.Filename))
	if !allowedExtensions[suffix] {
		httpx.WriteError(w, http.StatusBadRequest,
			"File type not allowed. Accepted: .pdf, .doc, .docx, .jpg, .jpeg, .png", "")
		return
	}
	if ct := header.Header.Get("Content-Type"); ct != "" && !allowedMimeTypes[ct] {
		httpx.WriteError(w, http.StatusBadRequest, "MIME type not allowed.", "")
		return
	}

	contents, err := io.ReadAll(io.LimitReader(file, config.S.UploadMaxBytes+1))
	if err != nil {
		httpx.WriteError(w, http.StatusBadRequest, "Could not read file.", "")
		return
	}
	if int64(len(contents)) > config.S.UploadMaxBytes {
		httpx.WriteError(w, http.StatusRequestEntityTooLarge,
			fmt.Sprintf("File exceeds %d MB limit.", config.S.UploadMaxBytes/(1024*1024)), "")
		return
	}
	if len(contents) == 0 {
		httpx.WriteError(w, http.StatusBadRequest, "Empty file.", "")
		return
	}

	dir := uploadDir()
	originalName := header.Filename
	if originalName == "" {
		originalName = "upload"
	}

	// TODO(go-port): the Python route kicks off an async AI document scan
	// (scan_document) that fills `scanned_text`. AI scanning is out of scope for
	// the Go port, so scanned_text is left nil ("[scan pending]").

	if existing, ok := getDoc(userKey, docType); ok {
		_ = os.Remove(filepath.Join(dir, existing.StoredName))
		existing.OriginalName = originalName
		existing.StoredName = newStoredName(suffix)
		existing.ScannedText = nil
		if err := os.WriteFile(filepath.Join(dir, existing.StoredName), contents, 0o644); err != nil {
			httpx.WriteError(w, http.StatusInternalServerError, "Could not store file.", httpx.CRV1006)
			return
		}
		db.DB.Save(&existing)
		metrics.Increment("doc_uploads")
		httpx.JSON(w, http.StatusOK, map[string]any{"ok": true, "original_name": existing.OriginalName})
		return
	}

	storedName := newStoredName(suffix)
	if err := os.WriteFile(filepath.Join(dir, storedName), contents, 0o644); err != nil {
		httpx.WriteError(w, http.StatusInternalServerError, "Could not store file.", httpx.CRV1006)
		return
	}
	doc := models.Document{
		UserKey:      userKey,
		DocType:      docType,
		OriginalName: originalName,
		StoredName:   storedName,
	}
	if err := db.DB.Create(&doc).Error; err != nil {
		httpx.WriteError(w, http.StatusInternalServerError, "Could not save document.", httpx.CRV1002)
		return
	}
	metrics.Increment("doc_uploads")
	httpx.JSON(w, http.StatusOK, map[string]any{"ok": true, "original_name": doc.OriginalName})
}

func remove(w http.ResponseWriter, r *http.Request) {
	sess := security.SessionFromContext(r.Context())
	docType := chi.URLParam(r, "docType")
	if !validDocTypes[docType] {
		httpx.WriteError(w, http.StatusBadRequest, "Invalid doc_type.", "")
		return
	}
	doc, ok := getDoc(sess.Sub, docType)
	if !ok {
		httpx.WriteError(w, http.StatusNotFound, "Document not found.", "")
		return
	}
	_ = os.Remove(filepath.Join(uploadDir(), doc.StoredName))
	db.DB.Delete(&doc)
	metrics.Increment("doc_deletes")
	httpx.JSON(w, http.StatusOK, map[string]any{"ok": true})
}

func download(w http.ResponseWriter, r *http.Request) {
	sess := security.SessionFromContext(r.Context())
	docType := chi.URLParam(r, "docType")
	if !validDocTypes[docType] {
		httpx.WriteError(w, http.StatusBadRequest, "Invalid doc_type.", "")
		return
	}
	doc, ok := getDoc(sess.Sub, docType)
	if !ok {
		httpx.WriteError(w, http.StatusNotFound, "Document not found.", "")
		return
	}

	dir, _ := filepath.Abs(uploadDir())
	filePath, _ := filepath.Abs(filepath.Join(dir, doc.StoredName))
	if !strings.HasPrefix(filePath, dir) {
		httpx.WriteError(w, http.StatusBadRequest, "Invalid file path.", "")
		return
	}
	if _, err := os.Stat(filePath); err != nil {
		httpx.WriteError(w, http.StatusNotFound, "File missing on server.", "")
		return
	}
	metrics.Increment("doc_downloads")
	w.Header().Set("Content-Type", "application/octet-stream")
	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%q", doc.OriginalName))
	http.ServeFile(w, r, filePath)
}

func newStoredName(suffix string) string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b) + suffix
}
